## Import Dependencies

In [1]:
import pandas as pd
import numpy as np
from lls import LinearLeastSquare
from train_test_split import train_test_split
from sklearn.metrics import mean_absolute_error


## Import the dataset

In [2]:
data = pd.read_csv('Dollar_Rial_Price_Dataset.csv')
data.head()

Unnamed: 0        Date Persian_Date    Open     Low    High   Close
0           0  11/27/2011   1390/09/06  13,700  13,700  13,700  13,700
1           1  11/28/2011   1390/09/07  13,440  13,440  13,440  13,440
2           2  11/29/2011   1390/09/08  13,350  13,350  13,350  13,350
3           3  11/30/2011   1390/09/09  13,400  13,400  13,400  13,400
4           4   12/1/2011   1390/09/10  13,500  13,500  13,500  13,500

#### Preprocessing on dataset

In [3]:
data.drop(['Unnamed: 0'], inplace=True, axis=1)
data.drop(['Persian_Date'], inplace=True, axis=1)
data.head()

Date    Open     Low    High   Close
0  11/27/2011  13,700  13,700  13,700  13,700
1  11/28/2011  13,440  13,440  13,440  13,440
2  11/29/2011  13,350  13,350  13,350  13,350
3  11/30/2011  13,400  13,400  13,400  13,400
4   12/1/2011  13,500  13,500  13,500  13,500

In [4]:
def my_func(x: str):
    return x.replace(',', '')


data.drop_duplicates(inplace=True)
data['High'] = data['High'].apply(my_func)
data['Low'] = data['Low'].apply(my_func)
data['Open'] = data['Open'].apply(my_func)
data['Close'] = data['Close'].apply(my_func)
data.head()

Date   Open    Low   High  Close
0  11/27/2011  13700  13700  13700  13700
1  11/28/2011  13440  13440  13440  13440
2  11/29/2011  13350  13350  13350  13350
3  11/30/2011  13400  13400  13400  13400
4   12/1/2011  13500  13500  13500  13500

In [5]:
data['High'] = pd.to_numeric(data['High'], errors='coerce')
data['Low'] = pd.to_numeric(data['Low'], errors='coerce')
data['Open'] = pd.to_numeric(data['Open'], errors='coerce')
data['Close'] = pd.to_numeric(data['Close'], errors='coerce')

## Divide dataset to Ahmadinejad, Rouhani and Raisi's presidency

In [6]:
data['Date'] = pd.to_datetime(data['Date'], format="%m/%d/%Y")
ahmadinejad_presidency = data.loc[(data['Date'] >= "11/27/2011") & (data['Date'] < "08/03/2013")].reset_index()
rohani_presidency = data.loc[(data['Date'] >= "08/03/2013") & (data['Date'] < "08/03/2021")].reset_index()
raisi_presidency = data.loc[(data['Date'] >= "08/03/2021") & (data['Date'] < "07/26/2023")].reset_index()

rohani_presidency.head()

index       Date   Open    Low   High  Close
0    482 2013-08-03  32100  32050  32900  32850
1    483 2013-08-04  32650  32400  32800  32400
2    484 2013-08-05  32150  31900  32200  32000
3    485 2013-08-06  31900  31750  31950  31900
4    486 2013-08-07  32050  32050  32350  32250

## Show the highest dollar price in Ahmadinejad, Rouhani and Raisi's presidency respectively

In [8]:
print(f'\nHighest Dollar Price of Ahmadinejad Presidency is {ahmadinejad_presidency["High"].max()} Rial.\n')
print(f'Highest Dollar Price of Rohani Presidency is {rohani_presidency["High"].max()} Rial.\n')
print(f'Highest Dollar Price of Raisi Presidency is {raisi_presidency["High"].max()} Rial.\n')


Highest Dollar Price of Ahmadinejad Presidency is 39700 Rial.

Highest Dollar Price of Rohani Presidency is 320060 Rial.

Highest Dollar Price of Raisi Presidency is 555600 Rial.



|  ------------Max Dollar Price In 3 Presidency----------------        |
|-------------------------------------------------------------------------|

|                  |     Price(Rial)    |
|------------------|------------|
|      Max Price Of Dollar in Ahmadinejad Presidency      | 39700 | 
|      Max Price Of Dollar in Rohani  Presidency           | 320060 |
|      Max Price Of Dollar in Raisi  Presidency             | 555600 |

## Show the lowest dollar price in Ahmadinejad, Rouhani and Raisi's presidency respectively

In [10]:
print(f'\nLowest Dollar Price of Ahmadinejad Presidency is {ahmadinejad_presidency["Low"].min()} Rial.\n')
print(f'Lowest Dollar Price of Rohani Presidency is {rohani_presidency["Low"].min()} Rial.\n')
print(f'Lowest Dollar Price of Raisi Presidency is {raisi_presidency["Low"].min()} Rial.\n')


Lowest Dollar Price of Ahmadinejad Presidency is 13350 Rial.

Lowest Dollar Price of Rohani Presidency is 12850 Rial.

Lowest Dollar Price of Raisi Presidency is 251250 Rial.



|  ------------Min Dollar Price In 3 Presidency----------------        |
|-------------------------------------------------------------------------|

|                  |     Price(Rial)    |
|------------------|------------|
|      Min Price Of Dollar in Ahmadinejad Presidency      | 13350  | 
|      Min Price Of Dollar in Rohani Presidency           | 12850  |
|      Min Price Of Dollar in Rohani Presidency            | 251250 |

## Split each dataset to train and test with train_test_split function

In [11]:
x_ahmadinejad = ahmadinejad_presidency['Date'].values.astype(np.int64)
y_ahmadinejad = ahmadinejad_presidency['High'].values

ah_x_train, ah_x_test, ah_y_train, ah_y_test = train_test_split(x_ahmadinejad, y_ahmadinejad)
ah_x_train, ah_x_test, ah_y_train, ah_y_test  = ah_x_train.reshape(-1, 1), ah_x_test.reshape(-1, 1), ah_y_train.reshape(-1, 1), ah_y_test.reshape(-1, 1)

In [12]:
x_rohani = rohani_presidency['Date'].values.astype(np.int64)
y_rohani = rohani_presidency['High'].values

ro_x_train, ro_x_test, ro_y_train, ro_y_test = train_test_split(x_rohani, y_rohani)
ro_x_train, ro_x_test, ro_y_train, ro_y_test  = ro_x_train.reshape(-1, 1), ro_x_test.reshape(-1, 1), ro_y_train.reshape(-1, 1), ro_y_test.reshape(-1, 1)

In [13]:
x_raisi = raisi_presidency['Date'].values.astype(np.int64)
y_raisi = raisi_presidency['High'].values


rai_x_train, rai_x_test, rai_y_train, rai_y_test = train_test_split(x_rohani, y_rohani)
rai_x_train, rai_x_test, rai_y_train, rai_y_test  = rai_x_train.reshape(-1, 1), rai_x_test.reshape(-1, 1), rai_y_train.reshape(-1, 1), rai_y_test.reshape(-1, 1)

## Fit the LLS model on each training dataset

In [14]:
lls_ah = LinearLeastSquare()
lls_ah.fit(ah_x_train, ah_y_train)
ah_y_pred = lls_ah.predict(ah_x_test)

In [15]:
lls_ro = LinearLeastSquare()
lls_ro.fit(ro_x_train, ro_y_train)
ro_y_pred = lls_ro.predict(ro_x_test)

In [16]:
lls_rai = LinearLeastSquare()
lls_rai.fit(rai_x_train, rai_y_train)
rai_y_pred = lls_rai.predict(rai_x_test)

## Evaluate each model on test dataset using MAE loss function in Scikit-Learn

In [18]:
print(f'\nMAE of Ahmadinejad is {mean_absolute_error(ah_y_test, ah_y_pred)}.\n')
print(f'MAE of Rohani is {mean_absolute_error(ro_y_test, ro_y_pred)}.\n')
print(f'MAE of Raisi is {mean_absolute_error(rai_y_test, rai_y_pred)}.\n')


MAE of Ahmadinejad is 2.0110322469723354e+24.

MAE of Rohani is 1.6516357696253647e+26.

MAE of Raisi is 1.498722684643541e+27.



| MAE for Dollar Price Prediction In 3 presidency       |
|-------------------------------------------------------------------------|

|                  |     Price(Rial)    |
|------------------|------------|
|      MAE for Ahmadinejad Presidency      |2.0110322469723354e+24| 
|      MAE for Rohani Presidency          | 1.6516357696253647e+26 |
|      MAE for Raisi Presidency            | 1.498722684643541e+27 |